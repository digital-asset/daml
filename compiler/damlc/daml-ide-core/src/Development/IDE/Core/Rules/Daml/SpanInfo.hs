-- Copyright (c) 2024 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
-- SPDX-License-Identifier: Apache-2.0

{-# LANGUAGE BlockArguments #-}
{-# LANGUAGE DeriveAnyClass #-}
{-# LANGUAGE DerivingStrategies #-}

module Development.IDE.Core.Rules.Daml.SpanInfo
    ( SpansInfo (..)
    , SpanInfo (..)
    , SpanType (..)
    , SpanSource (..)
    , SpanDoc (..)
    , getSpanInfo
    ) where

import Data.Aeson (FromJSON, ToJSON)
import Data.Text (Text)
import Development.IDE.Core.Shake (use_)
import Development.IDE.Types.Location (NormalizedFilePath)
import Development.Shake (Action)
import GHC.Generics (Generic)

import "ghc-lib-parser" DynFlags (DynFlags)
import "ghc-lib-parser" Outputable (showPpr)
import "ghc-lib-parser" TyCoRep (Type)

import qualified Data.Text as T
import qualified Development.IDE.Core.RuleTypes as IDE
import qualified Development.IDE.Spans.Common as IDE
import qualified Development.IDE.Spans.Type as IDE

getSpanInfo :: DynFlags -> NormalizedFilePath -> Action SpansInfo
getSpanInfo dflags inputFile =
  convertSpansInfo dflags <$> use_ IDE.GetSpanInfo inputFile

convertSpansInfo :: DynFlags -> IDE.SpansInfo -> SpansInfo
convertSpansInfo dflags = convertSpansInfo'
  where
    convertSpansInfo' :: IDE.SpansInfo -> SpansInfo
    convertSpansInfo' spansInfo = SpansInfo
      { expressions = convertSpanInfo <$> IDE.spansExprs spansInfo
      , constraints = convertSpanInfo <$> IDE.spansConstraints spansInfo
      }

    convertSpanInfo :: IDE.SpanInfo -> SpanInfo
    convertSpanInfo spanInfo = SpanInfo
      { startLine = IDE.spaninfoStartLine spanInfo
      , startCol = IDE.spaninfoStartCol spanInfo
      , endLine = IDE.spaninfoEndLine spanInfo
      , endCol = IDE.spaninfoEndCol spanInfo
      , type_ = convertSpanType (IDE.spaninfoType spanInfo)
      , source = convertSpanSource (IDE.spaninfoSource spanInfo)
      , doc = convertSpanDoc (IDE.spaninfoDocs spanInfo)
      }

    convertSpanType :: Maybe Type -> SpanType
    convertSpanType = SpanType . \case
      Nothing -> "<missing>"
      Just ty -> T.pack (showPpr dflags ty)

    convertSpanSource :: IDE.SpanSource -> SpanSource
    convertSpanSource = \case
      IDE.Named name -> Named (T.pack (showPpr dflags name))
      IDE.SpanS span -> Span (T.pack (showPpr dflags span))
      IDE.Lit lit -> Lit (T.pack lit)
      IDE.NoSource -> NoSource

    convertSpanDoc :: IDE.SpanDoc -> SpanDoc
    convertSpanDoc = SpanDoc . \case
      IDE.SpanDocString ds -> T.lines (T.pack (showPpr dflags ds))
      IDE.SpanDocText ts -> ts

data SpansInfo = SpansInfo
  { expressions :: ![SpanInfo]
  , constraints :: ![SpanInfo]
  }
  deriving stock (Eq, Ord, Show, Generic)
  deriving anyclass (FromJSON, ToJSON)

data SpanInfo = SpanInfo
  { startLine :: !Int
  , startCol :: !Int
  , endLine :: !Int
  , endCol :: !Int
  , type_ :: !SpanType
  , source :: !SpanSource
  , doc :: !SpanDoc
  }
  deriving stock (Eq, Ord, Show, Generic)
  deriving anyclass (FromJSON, ToJSON)

newtype SpanType = SpanType Text
  deriving stock (Eq, Ord, Show, Generic)
  deriving newtype (FromJSON, ToJSON)

data SpanSource
  = Named !Text
  | Span !Text
  | Lit !Text
  | NoSource
  deriving stock (Eq, Ord, Show, Generic)
  deriving anyclass (FromJSON, ToJSON)

newtype SpanDoc = SpanDoc [Text]
  deriving stock (Eq, Ord, Show, Generic)
  deriving newtype (FromJSON, ToJSON)
