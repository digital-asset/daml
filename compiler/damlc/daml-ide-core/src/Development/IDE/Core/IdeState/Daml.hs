-- Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
-- SPDX-License-Identifier: Apache-2.0

module Development.IDE.Core.IdeState.Daml
    ( IdeState
    , getDamlIdeState
    , withDamlIdeState
    , enabledPlugins
    ) where

import Data.Default
import qualified Language.Haskell.LSP.Messages as LSP

import Control.Exception
import DA.Daml.Options
import DA.Daml.Options.Types
import qualified DA.Service.Logger as Logger
import qualified DA.Daml.LF.ScenarioServiceClient as Scenario
import Development.IDE.Core.Debouncer
import Development.IDE.Core.API
import Development.IDE.Core.Rules.Daml
import Development.IDE.Plugin
import Development.IDE.Plugin.Completions as Completions
import Development.IDE.Plugin.CodeAction as CodeAction
import qualified Development.IDE.Types.Logger as IdeLogger
import Development.IDE.Types.Options
import qualified Language.Haskell.LSP.Types as LSP
import qualified Language.Haskell.LSP.Types.Capabilities as LSP

getDamlIdeState
    :: Options
    -> Maybe Scenario.Handle
    -> Logger.Handle IO
    -> Debouncer LSP.NormalizedUri
    -> LSP.ClientCapabilities
    -> IO LSP.LspId
    -> (LSP.FromServerMessage -> IO ())
    -> VFSHandle
    -> IdeReportProgress
    -> IO IdeState
getDamlIdeState compilerOpts mbScenarioService loggerH debouncer caps getLspId eventHandler vfs reportProgress = do
    let rule = mainRule compilerOpts <> pluginRules enabledPlugins
    damlEnv <- mkDamlEnv compilerOpts mbScenarioService
    initialise caps rule getLspId eventHandler (toIdeLogger loggerH) debouncer damlEnv (toCompileOpts compilerOpts reportProgress) vfs

enabledPlugins :: Plugin a
enabledPlugins = Completions.plugin <> CodeAction.plugin

-- Wrapper for the common case where the scenario service
-- will be started automatically (if enabled)
-- and we use the builtin VFSHandle. We always disable
-- the debouncer here since this is not used in the IDE.
withDamlIdeState
    :: Options
    -> Logger.Handle IO
    -> (LSP.FromServerMessage -> IO ())
    -> (IdeState -> IO a)
    -> IO a
withDamlIdeState opts@Options{..} loggerH eventHandler f = do
    scenarioServiceConfig <- Scenario.readScenarioServiceConfig
    Scenario.withScenarioService' optScenarioService loggerH scenarioServiceConfig $ \mbScenarioService -> do
        vfs <- makeVFSHandle
        -- We only use withDamlIdeState outside of the IDE where we do not care about
        -- progress reporting.
        bracket
            (getDamlIdeState opts mbScenarioService loggerH noopDebouncer def (pure $ LSP.IdInt 0) eventHandler vfs (IdeReportProgress False))
            shutdown
            f

-- | Adapter to the IDE logger module.
toIdeLogger :: Logger.Handle IO -> IdeLogger.Logger
toIdeLogger h = IdeLogger.Logger $ \case
    IdeLogger.Error -> Logger.logError h
    IdeLogger.Warning -> Logger.logWarning h
    IdeLogger.Info -> Logger.logInfo h
    IdeLogger.Debug -> Logger.logDebug h
    IdeLogger.Telemetry -> Logger.logTelemetry h
