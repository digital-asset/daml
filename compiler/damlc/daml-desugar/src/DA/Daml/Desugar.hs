-- Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
-- SPDX-License-Identifier: Apache-2.0

module DA.Daml.Desugar (desugar) where

import DA.Daml.Compiler.Output (diagnosticsLogger)
import DA.Daml.Options.Types (Options, getLogger)
import Data.Text (Text)
import Development.IDE.Core.API (runActionSync, setFilesOfInterest)
import Development.IDE.Core.IdeState.Daml (withDamlIdeState)
import Development.IDE.Core.Rules (GetParsedModule (..))
import Development.IDE.Core.RuleTypes (GhcSession (..))
import Development.IDE.Core.Shake (use_)
import Development.IDE.GHC.Util (hscEnv)
import Development.IDE.Types.Location (toNormalizedFilePath')

import qualified Data.HashSet as HashSet
import qualified Data.Text as T

import "ghc-lib" GHC (pm_parsed_source)
import "ghc-lib-parser" HscTypes (hsc_dflags)

import qualified "ghc-lib-parser" Outputable as GHC

desugar :: Options -> FilePath -> IO Text
desugar opts inputFile = do
  loggerH <- getLogger opts "daml-desugar"
  inputFile <- pure $ toNormalizedFilePath' inputFile
  withDamlIdeState opts loggerH diagnosticsLogger $ \ide -> do
    setFilesOfInterest ide (HashSet.singleton inputFile)
    runActionSync ide $ do
      dflags <- hsc_dflags . hscEnv <$> use_ GhcSession inputFile
      parsed <- pm_parsed_source <$> use_ GetParsedModule inputFile
      pure . ensureNewlineAtEof . T.pack . GHC.showSDoc dflags . GHC.ppr $ parsed

ensureNewlineAtEof :: Text -> Text
ensureNewlineAtEof t
  | Just (_, '\n') <- T.unsnoc t = t
  | otherwise                    = t <> "\n"
