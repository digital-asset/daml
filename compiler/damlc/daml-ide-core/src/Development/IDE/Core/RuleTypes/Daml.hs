-- Copyright (c) 2019 The DAML Authors. All rights reserved.
-- SPDX-License-Identifier: Apache-2.0

{-# OPTIONS_GHC -Wno-orphans #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE TypeFamilies #-}

-- | This module extends the general Haskell rule types in
-- `Development.IDE.Core.RuleTypes` with DAML specific rule types
-- such as those for producing DAML LF.
module Development.IDE.Core.RuleTypes.Daml(
    module Development.IDE.Core.RuleTypes,
    module Development.IDE.Core.RuleTypes.Daml
    ) where

import Control.DeepSeq
import Data.Binary
import qualified Data.ByteString as BS
import Data.Hashable
import Data.Map.Strict (Map)
import Data.Set (Set)
import Data.Typeable (Typeable)
import Development.Shake
import GHC.Generics (Generic)
import "ghc-lib-parser" Module (UnitId)
import Development.IDE.Core.Service.Daml

import Development.IDE.Types.Location
import Development.IDE.Core.RuleTypes

import DA.Daml.DocTest
import qualified DA.Daml.LF.Ast as LF
import qualified DA.Daml.LF.ScenarioServiceClient as SS

import Language.Haskell.HLint4

type instance RuleResult GenerateDalf = LF.Module
type instance RuleResult GenerateRawDalf = LF.Module

-- | A newtype wrapper for LF.Package that does not force the modules
-- in the package to be evaluated to NF. This is useful since we
-- already force the evaluation when we build the modules
-- and avoids having to traverse all dependencies of a module
-- if only that module changed.
newtype WhnfPackage = WhnfPackage { getWhnfPackage :: LF.Package }
    deriving Show

instance NFData WhnfPackage where
    rnf (WhnfPackage (LF.Package ver modules)) =
        modules `seq` rnf ver

type instance RuleResult GeneratePackage = WhnfPackage
type instance RuleResult GenerateRawPackage = WhnfPackage
type instance RuleResult GeneratePackageDeps = WhnfPackage

data DalfPackage = DalfPackage
    { dalfPackageId :: LF.PackageId
    , dalfPackagePkg :: LF.ExternalPackage
    , dalfPackageBytes :: BS.ByteString
    } deriving (Show, Eq, Generic)

instance NFData DalfPackage

type instance RuleResult GeneratePackageMap = Map UnitId DalfPackage

-- | Runs all scenarios in the given file (but not scenarios in imports).
type instance RuleResult RunScenarios = [(VirtualResource, Either SS.Error SS.ScenarioResult)]

-- | Encode a module and produce a hash of the module and all its transitive dependencies.
-- The hash is used to decide if a module needs to be reloaded in the scenario service.
type instance RuleResult EncodeModule = (SS.Hash, BS.ByteString)

-- | Create a scenario context for a given module. This context is valid both for the module
-- itself but also for all of its transitive dependencies.
type instance RuleResult CreateScenarioContext = SS.ContextId

-- ^ A map from a file A to a file B whose scenario context should be
-- used for executing scenarios in A. We use this when running the scenarios
-- in transitive dependencies of the files of interest so that we only need
-- one scenario context per file of interest.
type instance RuleResult GetScenarioRoots = Map NormalizedFilePath NormalizedFilePath

-- ^ The root for the given file based on GetScenarioRoots.
-- This is a separate rule so we can avoid rerunning scenarios if
-- only the roots of other files have changed.
type instance RuleResult GetScenarioRoot = NormalizedFilePath

-- | These rules manage access to the global state in
-- envOfInterestVar and envOpenVirtualResources.
type instance RuleResult GetOpenVirtualResources = Set VirtualResource

data GenerateDalf = GenerateDalf
    deriving (Eq, Show, Typeable, Generic)
instance Binary   GenerateDalf
instance Hashable GenerateDalf
instance NFData   GenerateDalf

data GenerateRawDalf = GenerateRawDalf
    deriving (Eq, Show, Typeable, Generic)
instance Binary   GenerateRawDalf
instance Hashable GenerateRawDalf
instance NFData   GenerateRawDalf

data GeneratePackage = GeneratePackage
    deriving (Eq, Show, Typeable, Generic)
instance Binary   GeneratePackage
instance Hashable GeneratePackage
instance NFData   GeneratePackage

data GenerateRawPackage = GenerateRawPackage
    deriving (Eq, Show, Typeable, Generic)
instance Binary   GenerateRawPackage
instance Hashable GenerateRawPackage
instance NFData   GenerateRawPackage

data GeneratePackageDeps = GeneratePackageDeps
    deriving (Eq, Show, Typeable, Generic)
instance Binary   GeneratePackageDeps
instance Hashable GeneratePackageDeps
instance NFData   GeneratePackageDeps

data GeneratePackageMap = GeneratePackageMap
    deriving (Eq, Show, Typeable, Generic)
instance Binary   GeneratePackageMap
instance Hashable GeneratePackageMap
instance NFData   GeneratePackageMap

data RunScenarios = RunScenarios
    deriving (Eq, Show, Typeable, Generic)
instance Binary   RunScenarios
instance Hashable RunScenarios
instance NFData   RunScenarios

data EncodeModule = EncodeModule
    deriving (Eq, Show, Typeable, Generic)
instance Binary   EncodeModule
instance Hashable EncodeModule
instance NFData   EncodeModule

data CreateScenarioContext = CreateScenarioContext
    deriving (Eq, Show, Typeable, Generic)
instance Binary   CreateScenarioContext
instance Hashable CreateScenarioContext
instance NFData   CreateScenarioContext

data GetScenarioRoots = GetScenarioRoots
    deriving (Eq, Show, Typeable, Generic)
instance Hashable GetScenarioRoots
instance NFData   GetScenarioRoots

data GetScenarioRoot = GetScenarioRoot
    deriving (Eq, Show, Typeable, Generic)
instance Hashable GetScenarioRoot
instance NFData   GetScenarioRoot

data GetOpenVirtualResources = GetOpenVirtualResources
    deriving (Eq, Show, Typeable, Generic)
instance Hashable GetOpenVirtualResources
instance NFData   GetOpenVirtualResources

data GetDlintSettings = GetDlintSettings
    deriving (Eq, Show, Typeable, Generic)
instance Hashable GetDlintSettings
instance NFData   GetDlintSettings
instance NFData Hint where rnf = rwhnf
instance NFData Classify where rnf = rwhnf
instance Show Hint where show = const "<hint>"

type instance RuleResult GetDlintSettings = ([Classify], Hint)

data GetDlintDiagnostics = GetDlintDiagnostics
    deriving (Eq, Show, Typeable, Generic)
instance Hashable GetDlintDiagnostics
instance NFData   GetDlintDiagnostics

type instance RuleResult GetDlintDiagnostics = ()

data GenerateDocTestModule = GenerateDocTestModule
    deriving (Eq, Show, Typeable, Generic)
instance Hashable GenerateDocTestModule
instance NFData   GenerateDocTestModule

-- | File path of the generated module
type instance RuleResult GenerateDocTestModule = GeneratedModule

-- | Kick off things
type instance RuleResult OfInterest = ()

data OfInterest = OfInterest
    deriving (Eq, Show, Typeable, Generic)
instance Hashable OfInterest
instance NFData   OfInterest
