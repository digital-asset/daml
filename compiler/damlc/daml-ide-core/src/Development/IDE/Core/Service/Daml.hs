-- Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
-- SPDX-License-Identifier: Apache-2.0

module Development.IDE.Core.Service.Daml(
    VirtualResource(..),
    DamlEnv(..),
    mkDamlEnv,
    getDamlServiceEnv,
    IdeState, initialise, shutdown,
    runAction,
    runActionSync,
    setFilesOfInterest, modifyFilesOfInterest, setOpenVirtualResources, modifyOpenVirtualResources,
    writeProfile,
    getDiagnostics, unsafeClearDiagnostics,
    ideLogger
    ) where

import Control.Concurrent.Extra
import Control.Monad
import Data.HashMap.Strict (HashMap)
import qualified Data.HashMap.Strict as HashMap
import Data.HashSet (HashSet)
import qualified Data.HashSet as HashSet
import qualified Data.Text as T
import Data.Tuple.Extra
import Development.Shake

import Development.IDE.Types.Logger
import Development.IDE.Core.Debouncer
import Development.IDE.Core.Service hiding (initialise)
import Development.IDE.Core.FileStore
import qualified Development.IDE.Core.Service as IDE
import Development.IDE.Core.OfInterest
import Development.IDE.Core.RuleTypes.Daml
import Development.IDE.Core.Shake
import Development.IDE.Types.Location
import Development.IDE.Types.Options
import qualified Language.Haskell.LSP.Types.Capabilities as LSP
import qualified Language.Haskell.LSP.Messages as LSP
import qualified Language.Haskell.LSP.Types as LSP

import DA.Daml.Options.Types
import qualified DA.Daml.LF.Ast as LF
import qualified DA.Daml.LF.ScenarioServiceClient as SS

data DamlEnv = DamlEnv
  { envScenarioService :: Maybe SS.Handle
  , envOpenVirtualResources :: Var (HashSet VirtualResource)
  , envScenarioContexts :: MVar (HashMap NormalizedFilePath SS.ContextId)
  -- ^ This is a map from the file for which the context was created to
  -- the context id. We use this to track which scenario contexts
  -- are active so that we can GC inactive scenarios.
  -- This should eventually go away and we should track scenario contexts
  -- in the same way that we track diagnostics.
  , envPreviousScenarioContexts :: MVar [SS.ContextId]
  -- ^ The scenario contexts we used as GC roots in the last iteration.
  -- This is used to avoid unnecessary GC calls.
  , envDamlLfVersion :: LF.Version
  , envSkipScenarioValidation :: SkipScenarioValidation
  , envIsGenerated :: Bool
  }

instance IsIdeGlobal DamlEnv

mkDamlEnv :: Options -> Maybe SS.Handle -> IO DamlEnv
mkDamlEnv opts scenarioService = do
    openVRsVar <- newVar HashSet.empty
    scenarioContextsVar <- newMVar HashMap.empty
    previousScenarioContextsVar <- newMVar []
    pure DamlEnv
        { envScenarioService = scenarioService
        , envOpenVirtualResources = openVRsVar
        , envScenarioContexts = scenarioContextsVar
        , envPreviousScenarioContexts = previousScenarioContextsVar
        , envDamlLfVersion = optDamlLfVersion opts
        , envSkipScenarioValidation = optSkipScenarioValidation opts
        , envIsGenerated = optIsGenerated opts
        }

getDamlServiceEnv :: Action DamlEnv
getDamlServiceEnv = getIdeGlobalAction

setOpenVirtualResources :: IdeState -> HashSet VirtualResource -> IO ()
setOpenVirtualResources state resources = modifyOpenVirtualResources state (const resources)

modifyOpenVirtualResources :: IdeState -> (HashSet VirtualResource -> HashSet VirtualResource) -> IO ()
modifyOpenVirtualResources state f = do
    DamlEnv{..} <- getIdeGlobalState state
    vrs <- modifyVar envOpenVirtualResources $ pure . dupe . f
    logDebug (ideLogger state) $ "Set vrs of interest to: " <> T.pack (show $ HashSet.toList vrs)
    void $ shakeRun state []

initialise
    :: LSP.ClientCapabilities
    -> Rules ()
    -> IO LSP.LspId
    -> (LSP.FromServerMessage -> IO ())
    -> Logger
    -> Debouncer NormalizedUri
    -> DamlEnv
    -> IdeOptions
    -> VFSHandle
    -> IO IdeState
initialise caps mainRule getLspId toDiags logger debouncer damlEnv options vfs =
    IDE.initialise
        caps
        (do addIdeGlobal damlEnv
            mainRule)
        getLspId
        toDiags
        logger
        debouncer
        options
        vfs
