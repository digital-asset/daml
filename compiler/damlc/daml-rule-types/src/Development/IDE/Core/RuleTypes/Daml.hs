-- Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
-- SPDX-License-Identifier: Apache-2.0

{-# OPTIONS_GHC -Wno-orphans #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE TypeFamilies #-}

-- | This module extends the general Haskell rule types in
-- `Development.IDE.Core.RuleTypes` with DAML specific rule types
-- such as those for producing DAML LF.
module Development.IDE.Core.RuleTypes.Daml(
    module Development.IDE.Core.RuleTypes,
    module Development.IDE.Core.RuleTypes.Daml
    ) where

import Control.DeepSeq
import Data.Binary
import qualified Data.ByteString as BS
import Data.Hashable
import Data.Map.Strict (Map)
import Data.HashSet (HashSet)
import qualified Data.Text as T
import Data.Typeable (Typeable)
import Development.Shake
import GHC.Generics (Generic)
import "ghc-lib-parser" Module (UnitId)
import Development.IDE.GHC.Util
import Development.IDE.Types.Diagnostics
import Development.IDE.Types.Location
import Development.IDE.Core.RuleTypes

import DA.Daml.DocTest
import qualified DA.Daml.LF.Ast as LF
import qualified DA.Daml.LF.ScenarioServiceClient as SS

import Language.Haskell.HLint4

import HscTypes (ModIface, ModSummary)

type instance RuleResult GenerateDalf = LF.Module
type instance RuleResult GenerateSerializedDalf = ()
type instance RuleResult GenerateRawDalf = LF.Module

-- | A newtype wrapper for LF.Package that does not force the modules
-- in the package to be evaluated to NF. This is useful since we
-- already force the evaluation when we build the modules
-- and avoids having to traverse all dependencies of a module
-- if only that module changed.
newtype WhnfPackage = WhnfPackage { getWhnfPackage :: LF.Package }
    deriving Show

instance NFData WhnfPackage where
    rnf (WhnfPackage (LF.Package ver modules metadata)) =
        modules `seq` rnf ver `seq` rnf metadata

type instance RuleResult GeneratePackage = WhnfPackage
type instance RuleResult GenerateRawPackage = WhnfPackage
type instance RuleResult GeneratePackageDeps = WhnfPackage

newtype GeneratePackageMapFun = GeneratePackageMapFun (FilePath -> Action ([FileDiagnostic], Map UnitId LF.DalfPackage))
instance Show GeneratePackageMapFun where show _ = "GeneratePackageMapFun"
instance NFData GeneratePackageMapFun where rnf !_ = ()
-- | This matches how GhcSessionIO is handled in ghcide.
type instance RuleResult GeneratePackageMapIO = GeneratePackageMapFun
type instance RuleResult GeneratePackageMap = Map UnitId LF.DalfPackage
type instance RuleResult GenerateStablePackages = Map (UnitId, LF.ModuleName) LF.DalfPackage

data DamlGhcSession = DamlGhcSession (Maybe NormalizedFilePath)
    deriving (Show, Eq, Generic)
instance Binary DamlGhcSession
instance Hashable DamlGhcSession
instance NFData DamlGhcSession
type instance RuleResult DamlGhcSession = HscEnvEq

-- | Virtual resources
data VirtualResource = VRScenario
    { vrScenarioFile :: !NormalizedFilePath
    , vrScenarioName :: !T.Text
    } deriving (Eq, Ord, Show, Generic)
    -- ^ VRScenario identifies a scenario in a given file.
    -- This virtual resource is associated with the HTML result of
    -- interpreting the corresponding scenario.

instance Hashable VirtualResource
instance NFData VirtualResource

-- | Runs all scenarios in the given file (but not scenarios in imports).
type instance RuleResult RunScenarios = [(VirtualResource, Either SS.Error SS.ScenarioResult)]

-- | Encode a module and produce a hash of the module and all its transitive dependencies.
-- The hash is used to decide if a module needs to be reloaded in the scenario service.
type instance RuleResult EncodeModule = (SS.Hash, BS.ByteString)

-- | Create a scenario context for a given module. This context is valid both for the module
-- itself but also for all of its transitive dependencies.
type instance RuleResult CreateScenarioContext = SS.ContextId

-- ^ A map from a file A to a file B whose scenario context should be
-- used for executing scenarios in A. We use this when running the scenarios
-- in transitive dependencies of the files of interest so that we only need
-- one scenario context per file of interest.
type instance RuleResult GetScenarioRoots = Map NormalizedFilePath NormalizedFilePath

-- ^ The root for the given file based on GetScenarioRoots.
-- This is a separate rule so we can avoid rerunning scenarios if
-- only the roots of other files have changed.
type instance RuleResult GetScenarioRoot = NormalizedFilePath

-- | These rules manage access to the global state in
-- envOfInterestVar and envOpenVirtualResources.
type instance RuleResult GetOpenVirtualResources = HashSet VirtualResource

-- | This is used for on-disk incremental builds
type instance RuleResult ReadSerializedDalf = LF.Module
-- | Read the interface and the summary. This rule has a cutoff on the ABI hash of the interface.
-- It is important to get the modsummary from this rule rather than depending on the parsed module.
-- Otherwise you will always recompile module B if B depends on A and A changed even if Aâ€™s ABI stayed
-- the same
type instance RuleResult ReadInterface = (ModSummary, ModIface)

instance Show ModIface where
    show _ = "<ModIface>"

instance NFData ModIface where
    rnf = rwhnf

data GenerateDalf = GenerateDalf
    deriving (Eq, Show, Typeable, Generic)
instance Binary   GenerateDalf
instance Hashable GenerateDalf
instance NFData   GenerateDalf

data GenerateSerializedDalf = GenerateSerializedDalf
    deriving (Eq, Show, Typeable, Generic)
instance Binary GenerateSerializedDalf
instance Hashable GenerateSerializedDalf
instance NFData GenerateSerializedDalf

data ReadSerializedDalf = ReadSerializedDalf
    deriving (Eq, Show, Typeable, Generic)
instance Binary   ReadSerializedDalf
instance Hashable ReadSerializedDalf
instance NFData   ReadSerializedDalf

data ReadInterface = ReadInterface
    deriving (Eq, Show, Typeable, Generic)
instance Binary   ReadInterface
instance Hashable ReadInterface
instance NFData   ReadInterface

data GenerateRawDalf = GenerateRawDalf
    deriving (Eq, Show, Typeable, Generic)
instance Binary   GenerateRawDalf
instance Hashable GenerateRawDalf
instance NFData   GenerateRawDalf

data GeneratePackage = GeneratePackage
    deriving (Eq, Show, Typeable, Generic)
instance Binary   GeneratePackage
instance Hashable GeneratePackage
instance NFData   GeneratePackage

data GenerateRawPackage = GenerateRawPackage
    deriving (Eq, Show, Typeable, Generic)
instance Binary   GenerateRawPackage
instance Hashable GenerateRawPackage
instance NFData   GenerateRawPackage

data GeneratePackageDeps = GeneratePackageDeps
    deriving (Eq, Show, Typeable, Generic)
instance Binary   GeneratePackageDeps
instance Hashable GeneratePackageDeps
instance NFData   GeneratePackageDeps

data GeneratePackageMap = GeneratePackageMap
    deriving (Eq, Show, Typeable, Generic)
instance Binary   GeneratePackageMap
instance Hashable GeneratePackageMap
instance NFData   GeneratePackageMap

data GeneratePackageMapIO = GeneratePackageMapIO deriving (Eq, Show, Typeable, Generic)
instance Hashable GeneratePackageMapIO
instance NFData   GeneratePackageMapIO
instance Binary   GeneratePackageMapIO

data GenerateStablePackages = GenerateStablePackages
   deriving (Eq, Show, Typeable, Generic)
instance Binary GenerateStablePackages
instance Hashable GenerateStablePackages
instance NFData GenerateStablePackages

data RunScenarios = RunScenarios
    deriving (Eq, Show, Typeable, Generic)
instance Binary   RunScenarios
instance Hashable RunScenarios
instance NFData   RunScenarios

data EncodeModule = EncodeModule
    deriving (Eq, Show, Typeable, Generic)
instance Binary   EncodeModule
instance Hashable EncodeModule
instance NFData   EncodeModule

data CreateScenarioContext = CreateScenarioContext
    deriving (Eq, Show, Typeable, Generic)
instance Binary   CreateScenarioContext
instance Hashable CreateScenarioContext
instance NFData   CreateScenarioContext

data GetScenarioRoots = GetScenarioRoots
    deriving (Eq, Show, Typeable, Generic)
instance Hashable GetScenarioRoots
instance NFData   GetScenarioRoots
instance Binary   GetScenarioRoots

data GetScenarioRoot = GetScenarioRoot
    deriving (Eq, Show, Typeable, Generic)
instance Hashable GetScenarioRoot
instance NFData   GetScenarioRoot
instance Binary   GetScenarioRoot

data GetOpenVirtualResources = GetOpenVirtualResources
    deriving (Eq, Show, Typeable, Generic)
instance Hashable GetOpenVirtualResources
instance NFData   GetOpenVirtualResources
instance Binary   GetOpenVirtualResources

data GetDlintSettings = GetDlintSettings
    deriving (Eq, Show, Typeable, Generic)
instance Hashable GetDlintSettings
instance NFData   GetDlintSettings
instance NFData Hint where rnf = rwhnf
instance NFData Classify where rnf = rwhnf
instance Show Hint where show = const "<hint>"
instance Binary GetDlintSettings

type instance RuleResult GetDlintSettings = ([Classify], Hint)

data GetDlintDiagnostics = GetDlintDiagnostics
    deriving (Eq, Show, Typeable, Generic)
instance Hashable GetDlintDiagnostics
instance NFData   GetDlintDiagnostics
instance Binary   GetDlintDiagnostics

type instance RuleResult GetDlintDiagnostics = ()

data GenerateDocTestModule = GenerateDocTestModule
    deriving (Eq, Show, Typeable, Generic)
instance Hashable GenerateDocTestModule
instance NFData   GenerateDocTestModule
instance Binary   GenerateDocTestModule

-- | File path of the generated module
type instance RuleResult GenerateDocTestModule = GeneratedModule

-- | Kick off things
type instance RuleResult OfInterest = ()

data OfInterest = OfInterest
    deriving (Eq, Show, Typeable, Generic)
instance Hashable OfInterest
instance NFData   OfInterest
instance Binary   OfInterest
