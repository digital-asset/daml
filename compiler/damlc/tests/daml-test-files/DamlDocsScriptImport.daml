-- | My documented module
module DamlDocsScriptImport where

import Daml.Script

-- | My documented script
main = script $ pure ()
