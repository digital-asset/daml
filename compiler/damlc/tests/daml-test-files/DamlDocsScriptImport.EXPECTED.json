{
  "md_adts": [],
  "md_anchor": "module-damldocsscriptimport-97896",
  "md_classes": [],
  "md_descr": "My documented module",
  "md_functions": [
    {
      "fct_anchor": "function-damldocsscriptimport-main-98373",
      "fct_context": [],
      "fct_descr": "My documented script",
      "fct_name": "main",
      "fct_type": {
        "TypeApp": [
          {
            "referenceAnchor": "type-daml-script-script-90248",
            "referencePackage": "daml-script-UNVERSIONED"
          },
          "Script",
          [
            {
              "TypeTuple": []
            }
          ]
        ]
      }
    }
  ],
  "md_instances": [],
  "md_interfaces": [],
  "md_name": "DamlDocsScriptImport",
  "md_templates": []
}
